capstone/src/test/java/com/google/sps/objects/CommentTest.java

```diff
@@ -0,0 +1,92 @@
+package com.google.sps.objects;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+import java.util.ArrayList;
+
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EmbeddedEntity;
+import com.google.gson.Gson;
+
+import com.google.appengine.tools.development.testing.LocalDatastoreServiceTestConfig;
+import com.google.appengine.tools.development.testing.LocalServiceTestHelper;
+import com.google.sps.Objects.Comment;
+
+/**
+ * Unit tests for Comment.
+ *
+ */
+public class CommentTest {
+
+  private static final String USER_ID = "123123123";
+  private static final String COMMENT_TEXT = "a great comment";
+  private static final long TIMESTAMP = 4324344;
+
+  private final LocalServiceTestHelper helper =
+      new LocalServiceTestHelper(
+          new LocalDatastoreServiceTestConfig()
+              .setDefaultHighRepJobPolicyUnappliedJobPercentage(0));
+
+  private Comment comment;
+
+  @Before
+  public void setUp() {
+    helper.setUp();
+    comment = 
+      new Comment(
+        TIMESTAMP, /* timestamp */ 
+        COMMENT_TEXT, /* commentText */ 
+        USER_ID /* userId */ );
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+    comment = null;
+  }
+
+  @Test
+  public void getCommentTextTest() {
+    assertEquals(comment.getCommentText(), COMMENT_TEXT);
+  }
+
+  @Test
+  public void getUserTest() {
+    assertEquals(comment.getUser(), USER_ID);
+  }
+
+  @Test
+  public void getCommentEntityTest() {
+    EmbeddedEntity entity = new EmbeddedEntity();
+    entity.setProperty("userId", USER_ID);
+    entity.setProperty("commentText", COMMENT_TEXT);
+    entity.setProperty("timestamp", TIMESTAMP);
+    Comment returnedComment = Comment.getCommentEntity(entity);
+
+    String jsonRetrieved = new Gson().toJson(returnedComment);
+    String jsonOriginal = new Gson().toJson(comment);
+    assertTrue(jsonRetrieved.equals(jsonOriginal));
+  }
+
+  @Test
+  public void toEntityTest() {
+    EmbeddedEntity entity = comment.toEntity(comment.getCommentText(), comment.getUser());
+    ArrayList<EmbeddedEntity> allComments = new ArrayList<>();
+    allComments.add(entity);
+    Entity commentEntity = new Entity("Comment");
+    commentEntity.setProperty("comments", allComments);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    datastore.put(commentEntity);
+
+    ArrayList<EmbeddedEntity> commentEntitys = (ArrayList<EmbeddedEntity>) commentEntity.getProperty("comments");
+    assertEquals(commentEntitys.get(0).getProperty("commentText"), COMMENT_TEXT);
+    assertEquals(commentEntitys.get(0).getProperty("userId"), USER_ID);
```

Comment:
I may be wrong and fundamentally misunderstand this test, so please don't be offended, but here are my take-aways:

1. toEntityTest calls toEntity on the very first line before configuring much about the test. I assume the comment is being setup in the `setUp`, but the structure of a test should be:

// Arrange - Setup the test as you need, show what inputs are required and are being tested. Anything outside of this scope can be thrown into `setUp`

// Act - ideally, 1 line. Run the thing you want to test.

// Assert - Given the inputs from arrange, assert anything that should be true as a side-effect/output.

2. Why are the two datastore lines in this test? You don't appear to be retrieving from the datastore.

3. Why are you putting the EmbeddedEntity into the comment entity as a property? In the assert block, from what I can tell, you are retrieving the same property that you set above, meaning you're mostly just testing the test, rather than some underlying feature.

My concrete suggestions are to move or even copy any arrange logic into this test to show what you're testing. This may just be replacing `comment.getCommentText()` with `COMMENT_TEXT` so I more clearly see you're testing that input matches output.

From there, you've already run the function that you're claiming you want to test in the test name, so I would just assert that the object maintained the values you input. It's possible you're also testing another feature or set of features with this unit test, in which case I would recommend additional unit tests named to identify what you're testing, and then using the 3 block structure so I can more clearly follow what the `// Act` is that you're testing.

Reply:
Looking back, I did this in the most roundabout way and basically just tested the test. I modified this to make a simple toEntity/fromEntity test to make sure the correct data was being stored. Thanks for explaining this!!